### ISSUE TITLE
Linux directory iteration may skip some directories on certain file systems

### ISSUE THREAD

**Issue by addemko** [darbyjohnston/DJV#381]
Hello!

When I use djv on certain virtual/'libfuse' based file systems under Linux, it omits directories. I've root caused the issue to the following code:

https://github.com/darbyjohnston/DJV/blob/master/lib/djvSystem/FileInfoFuncUnix.cpp#L57

Basically, for this file system directories always have a d_type is of DT_UNKNOWN (and never DT_DIR) which according to https://man7.org/linux/man-pages/man3/readdir.3.html is valid. The djv code assumes it will never be DT_UNKNOWN.

Ideally, it should also check for DT_UNKNOWN (and do a full stat check via File::Info::GetType) in addition to checking for DT_DIR.

Thanks!

**Comment by darbyjohnston** [comment]
Hi! Thanks for the bug report and troubleshooting. You mention that this is only on certain libfuse file systems? Is there a way to recreate one of these file systems for testing?

**Comment by addemko** [comment]
I discovered it through testing my own libfuse virtual file system, which is not yet released. I imagine it'd apply to most of the demo examples though ( https://github.com/libfuse/libfuse/tree/master/example ). It's by no means limited to fuse file systems though, and I wouldn't be surprised if other 'real' file systems also don't return a proper d_type. See the blurb in readdir:

```
         Currently, only some filesystems (among them: Btrfs, ext2,
              ext3, and ext4) have full support for returning the file
              type in d_type.  All applications must properly handle a
              return of DT_UNKNOWN.
```

Using a loop back block device is another way to try the other file systems.

Or one could just do a 'blind' fix... ie don't check d_type at all and do a File::Info for each file, verify that works and then come back and add the DT_DIR logic as an optimization 😄

**Comment by darbyjohnston** [comment]
Hi, I just checked in a 'blind' fix that should hopefully work. :) Each djv::System::File::Info is already getting a full stat in the constructor, so I replaced DT_DIR with checking the type, like:

`
info.getType() != System::File::Type::Directory
`

If you have a chance to build a local version from source and test the change that would be great. I just checked in changes to the build system a couple days ago which should make the process easier using a CMake super build. Note also that there are a bunch of WIP changes so the UI will look a bit different from the currently released version...

**Comment by addemko** [comment]
Looks good, thanks!

I tried commit 83a1b59ddab2bb09767c9a73d8ec38dd6af0d1c8 which works (as it includes your 44fafce49010fe73c75272c11044b5629a3c4a73 fix) where as 829a18215803fa9dbdb89a20457950760d3db5e4 didn't work for me.

I think if you wanted you could use the `d_name` and `d_type` from the `dirent` and only fall back to creating an `Info` if its an DT_UNKNOWN, but maybe saving a bunch of `stat` calls is an overoptimization for these use cases. It looks good and works as is 👍

**Comment by darbyjohnston** [comment]
Great, thanks for testing it out, I'll include this change with the next release.

It would have been nice to avoid the stat for each file since it is a lot of overhead, but I wanted to show the file information in the file browser so it's necessary. I had been thinking of adding an option in the file browser UI to toggle the behavior, sort of a "fast no info" or "slow and lot's of info" option, but thought that might be too complicated.
